File: 54-SpiralMatrix/54-SpiralMatrix.java
// Last updated: 8/2/2025, 4:04:40 PM
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int minc = 0;
        int maxc = matrix[0].length - 1;
        int minr = 0;
        int maxr = matrix.length - 1;
        List<Integer> result = new ArrayList<>();
        int totalElements = matrix.length * matrix[0].length;
        int count = 0;
        
        while (count < totalElements) {
            // Traverse from left to right
            for (int i = minc; i <= maxc && count < totalElements; i++) {
                result.add(matrix[minr][i]);
                count++;
            }
            minr++;
            
            // Traverse downwards
            for (int j = minr; j <= maxr && count < totalElements; j++) {
                result.add(matrix[j][maxc]);
                count++;
            }
            maxc--;
            
            // Traverse from right to left
            for (int i = maxc; i >= minc && count < totalElements; i--) {
                result.add(matrix[maxr][i]);
                count++;
            }
            maxr--;
            
            // Traverse upwards
            for (int j = maxr; j >= minr && count < totalElements; j--) {
                result.add(matrix[j][minc]);
                count++;
            }
            minc++;
        }
        
        return result;
    }
}

File: 54-SpiralMatrix/54-SpiralMatrix.py
# Last updated: 8/2/2025, 4:04:44 PM
class Solution(object):
    def spiralOrder(self, ma):
        ans=[]
        m,n=len(ma),len(ma[0])
        t,b,l,r = 0,m-1,0,n-1
        while t<=b and l<=r:
            for i in range(l,r+1):
                ans.append(ma[t][i])
            t+=1
            if t>b:
                break
            for i in range(t,b+1):
                ans.append(ma[i][r])
            r-=1
            if r<l:
                break
            for i in range(r,l-1,-1):
                ans.append(ma[b][i])
            b-=1
            for i in range(b,t-1,-1):
                ans.append(ma[i][l])
            l+=1
        return ans
        
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        